### PR TITLE
[flang][OpenMP] Parse DOACROSS clause

### PR DESCRIPTION
**Pull request by kparzysz**
Extract the SINK/SOURCE parse tree elements into a separate class `OmpDoacross`, share them between DEPEND and DOACROSS clauses. Most of the changes in Semantics are to accommodate the new contents of OmpDependClause, and a mere introduction of OmpDoacrossClause.

There are no semantic checks specifically for DOACROSS.

### CODE REVIEW

**Comment by kparzysz** [comment]
The successor PR: https://github.com/llvm/llvm-project/pull/115397

**Review by kiranchandramohan — APPROVED** [review]
LG. Thanks Kryztof.

File: flang/lib/Semantics/resolve-directives.cpp [hunk 1]
```diff
@@ -1138,7 +1138,7 @@ bool AccAttributeVisitor::Pre(const parser::OpenACCCombinedConstruct &x) {
 static bool IsLastNameArray(const parser::Designator &designator) {
   const auto &name{GetLastName(designator)};
   const evaluate::DataRef dataRef{*(name.symbol)};
-  return common::visit(
+  return common::visit( //
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Semantics/resolve-directives.cpp:1141]: Yes, but it looks like a leftover from some wip changes.  Nothing else is changed in this function in the PR, so I'll remove it.

**Comment by kparzysz** [comment]
The code looked the same (except for the //) in the committed version, but when I ran clang-format on it, it reformatted this function.  If I added a commit that simply deleted the extra comment, the pre-commit clang-format would complain about formatting of this function, so I squashed/force-pushed the PR to make it look like this function wasn't modified at all.  Hopefully clang-format will leave it alone.

**Pull request merged**